### ISSUE TITLE
aws-ecs: FargateServiceProps Missing Field `forceNewDeployment`?

### ISSUE THREAD

**Issue by saidkharboutli** [aws/aws-cdk#27762]
### Describe the bug

Hello,

I just noticed that my CDK script has been doing every step of my NodeJS backend deployment except for restarting my ECS service upon ECR image update. There seems to be numerous threads on this specific issue, and the common answer is to use `forceNewDeployment`, which can be set from the CLI as well as within the Fargate service definition.

Currently, I am able to go into the AWS dashboard and manually check `forceNewDeployment` but that isn't the programmatic solution I am looking for. Also, once AWS spins up the new service and removes the old one, I have to check the option again for the next deployment.

The real underlying issue is that, for some reason, the [FargateServiceProps interface](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.FargateServiceProps.html) has no mention of this specific option. I am really not sure why, given that this issue of updating the ECS service after an ECR update is pretty wide-spread.

### Expected Behavior

I should be able to set the `forceNewDeployment` to true for the service when it is first built and thereafter. I really just want the ECS service to restart without downtime when the ECR image updates. If there is another way to do this that I have missed, please do let me know.

### Current Behavior

`forceNewDeployment` is not included as an option when creating an ECS service with CDK. Even checking it manually in the console does not stick, as the current service will spin down after the new one (that has the option set to False by default) is spun up.

### Reproduction Steps

Not certain if there are reproduction steps--it moreso a lack of an important field within the service definition.

### Possible Solution

Allow `forceNewDeployment` to be set to true for the creation of every service, forcing new ECS services to take the place of old ones when a new image is pushed to ECR.

### CDK CLI Version

2.103.1

### Node.js Version

20.9.0

### OS

Ubuntu

### Language

TypeScript

**Comment by daschaa** [comment]
@saidkharboutli Can you add a source on where you found the property `forceNewDeployment` ?
I had a look at the CloudFormation documentation, but I could not find it: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-service.html

**Comment by saidkharboutli** [comment]
@daschaa Upon further investigation, you do seem to be right--it's only a property in Terraform, and it's under another name (not camelCase).

Though regardless of its availability in CDK/Terraform now, if it's an option available in the AWS console, should we not be able to set it programmatically? There may be a reason it's not available, but it would be useful for the many people that have posted about this exact issue.

**Comment by daschaa** [comment]
@saidkharboutli I totally agree with you that the goal should be to cover the functionalities which you also have in the AWS console. However, if the field is not available in CloudFormation, we need to wait until CloudFormation implemented this.
I guess you could technically also implement a workaround with a custom resource and the AWS SDK, but I think the maintainers of the CDK are hesitant to implement new custom resources.

I would suggest to put a flag `needs-cfn` to this issue and leave it open.

**Comment by khushail** [comment]
Thanks @saidkharboutli  for reporting this.
Thanks @daschaa for your investigation. 

It is not supported by cloudformation yet. However one can submit a FR for this to be added on the [Cloudformation coverage roadmap](https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues).

**Comment by saidkharboutli** [comment]
@daschaa I appreciate your time on this endeavor greatly! I will likely take this over to the Cfn team then, and then perhaps once it is ready, this can be my first PR on the CDK project.

Thanks @khushail, I'll make that FR before I leave work today.

**Comment by rgoltz** [comment]
I'm adding the reference for the FR on CFN-github repo here: https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/1776 - Please up-vote this linked issue, to make CFN team aware of this.

**Comment by gem2578** [comment]
Hi it looks like CloudFormation now has the property https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-ecs-service-forcenewdeployment.html

**Comment by gem2578** [comment]
There is the property `ForceNewDeploymentNonce` now, so I was thinking if you set it to the image hash then it will only do a deployment when the hash change and that would only happen if there is a new image.

**Comment by yasomaru** [comment]
Since CloudFormation now supports the `ForceNewDeployment` property (as @gem2578 pointed out), I've submitted a PR to add L2 construct support for this feature.
#36797

**Comment by rgoltz** [comment]
Please mind that PR https://github.com/aws/aws-cdk/pull/35726 exists as well. It's already in review. Maybe there is some overlapping with your PR.

**Comment by yasomaru** [comment]
Thanks for pointing that out! I wasn't aware of #35726.

  I've reviewed the differences between our implementations:

  | Aspect | #35726 | #36797  |
  |--------|--------|---------------|
  | API | Method: `service.forceNewDeployment(true)` | Property: `forceNewDeployment: { enabled, nonce }` |
  | Nonce | Auto-generated timestamp | User-configurable |
  | Integration test | Not included | Included (with fixed nonce) |

  I think the property-based approach offers more flexibility since users can:
  - Use a stable nonce for controlled deployments
  - Use a time-varying value like `Date.now().toString()` for every-synth deployments

  However, I'm happy to close my PR if the maintainers prefer the method-based approach in #35726, or I can collaborate with @kg-aws to combine the best of both
  implementations.

  What would you recommend?

**Comment by gem2578** [comment]
Nonce is not required so is providing an auto-generated timestamp doing anything?

If the CDK is deploying multiple components it would be useful to have access to the nonce so you can set it to only forced to restart if the docker tag or hash change.

**Comment by yasomaru** [comment]
@gem2578 
Thanks for the feedback!

  Looking at the [CloudFormation
  documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-ecs-service-forcenewdeployment.html)

  > The value must be a unique, time-varying value like a timestamp, random string, or sequence number.

  AWS recommends using a timestamp, which triggers a new deployment on every `cdk deploy`. This makes sense for the typical use case where you're using a mutable
   tag like `latest` and always want to pull the newest image.

  However, as you mentioned, there are cases where you might want more control - for example, only triggering a deployment when the image digest actually
  changes. My implementation allows users to choose
  ```ts
  // AWS-recommended: deploy every time
  forceNewDeployment: {
    enabled: true,
    nonce: Date.now().toString(),
  }

  // Alternative: deploy only when image changes (if tracking hash externally)
  forceNewDeployment: {
    enabled: true,
    nonce: imageDigest,
  }
```
  This gives users flexibility while still supporting the default timestamp approach that AWS recommends.

**Comment by gem2578** [comment]
From the documentation it also says 

> Required: No

Which I think, implies that not providing the nonce is the same as providing a new value every time.

**Comment by yasomaru** [comment]
Thank you for raising that point about `Required: No` 
It was a great question that I hadn't considered.

  I ran some tests to verify the behavior

  **Test 1: No nonce provided**
  - 1st deploy: Stack created
  - 2nd deploy: `(no changes)` - no new deployment triggered

  **Test 2: With user-specified nonce**
  - 1st deploy with `sha256:abc123`: Stack created
  - 2nd deploy with `sha256:abc123` (same) : `(no changes)`
  - 3rd deploy with `sha256:xyz789` (changed): New ECS deployment triggered 

  Based on these results, it appears that
  - When no nonce is provided, CloudFormation doesn't auto-generate one, so
  subsequent deploys see no changes
  - The nonce value change is what signals CloudFormation to trigger a new
  deployment

  This means a nonce is actually necessary if you want to trigger deployments.
  Both approaches work:
  - **Timestamp nonce**: Deploy every time (aligns with AWS docs recommendation
  for `latest` tag use case)
  - **Hash/custom nonce**: Deploy only when the value changes (enables
  conditional deployments as you suggested)

  Thanks for the discussion, it helped clarify the expected behavior!

**Comment by gem2578** [comment]
@yasomaru thank you for testing that.

So it conditional, when enabled then it is required
